pkg/controller/job/job_controller.go

```diff
@@ -1089,7 +1101,7 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 				needsFlush = true
 				uncountedStatus.Succeeded = append(uncountedStatus.Succeeded, pod.UID)
 			}
-		} else if considerPodFailed || jobCtx.finishedCondition != nil {
+		} else if considerPodFailed || (jobCtx.finishedCondition != nil && !isSuccessCriteriaMetCondition(jobCtx.finishedCondition)) {
```

Comment:
Can we have the analogous `isFailureCondition` used here?
```suggestion
		} else if considerPodFailed || isFailureCondition(jobCtx.finishedCondition)) {
```
proposal, I think this will improve readability by long-inline conditions combining `&&` and `||` in one line.

Reply:
Sure, I'll try it.
Once we introduce the `isFailureTargetCondition`, I believe that we can replace complex conditional in some places.

Reply:
@mimowo I tried your proposal, but I found that the `jobCtx.finishedCondition != nil && 
 !isSuccessCriteriaMetCondition ` doesn't mean the `isFailureCondition`.

That conditional includes the "Complete", "Failed" as well. So, I guess that we need to keep the current conditional.

Reply:
Hm, but the intention is to check if the job is doomed to fail. IIUC isFailureCondition would check if this is FailureTarget or Failed. I think being explicit, rather than by negation is preferred

Reply:
>@mimowo I tried your proposal, but I found that the jobCtx.finishedCondition != nil &&  !isSuccessCriteriaMetCondition  doesn't mean the isFailureCondition.

Right, because `Complete` is neither SuccessCriteriaMet, not FailureTarget nor Failed.

So, we should not enter on `Complete` actually. I think this might be currently as subtle bug.

I think the intention of entering this branch of code is to only enter on failure. It is better to express that directly, than by negating success conditions.

I think we want to have `isFailureCondition`, where the function checks `FailureTarget || Failed`.

Reply:
> However, we should not enter this branch on Complete actually

We should. If we have reached succeeded==completions, the Job is declared completed, and any still running Pod should be deleted.

Reply:
However, if we have reached the successCriteria, the running Pods should be considered Succeeded or Failed, when they finish.

Reply:
> We should. If we have reached succeeded==completions, the Job is declared completed, and any still running Pod should be deleted and considered Failed.
> However, if we have reached the successCriteria, the running Pods should be considered Succeeded or Failed, when they finish.

Ah, I see it. But it is tricky, maybe leave a comment like this in this branch of code.

Reply:
> However, if we have reached the successCriteria, the running Pods should be considered Succeeded or Failed, when they finish.

I'm ok going with the current implementation for Alpha, but we need to revisit this line in Beta.

For this feature and JobPodReplacementPolicy, we probably should not declare the Pods as Failed right away, but wait for them to actually finish. It could even be considered a bug for JobPodReplacementPolicy, which is in Beta already.

cc @kannon92

Reply:
cc @dejanzele 

Could you verify the behavior and possibly open a bug?

Reply:
@alculquicondor @mimowo In my current approach, after the Job met the SuccessPolicy, still running pods aren't counted neither Succeeded nor Failed, as you can see here: https://github.com/kubernetes/kubernetes/blob/24123d1972d71b2297c6efb6b7bb55e46071f809/pkg/controller/job/job_controller_test.go#L3626-L3673

Running pods are recognized as abandoned.

Reply:
Thanks.
We still need to make sure that the Pod doesn't have a finalizer.

Also, I would expect the Pod to make it to "terminating" first, before it has reached a Failed/Succceeded phase.

Reply:
I will check

Reply:
> However, if we have reached the successCriteria, the running Pods should be considered Succeeded or Failed, when they finish.
> 
> I'm ok going with the current implementation for Alpha, but we need to revisit this line in Beta.
> 
> For this feature and JobPodReplacementPolicy, we probably should not declare the Pods as Failed right away, but wait for them to actually finish. It could even be considered a bug for JobPodReplacementPolicy, which is in Beta already.

I will update the future work section in the KEP to track this.